### DISCUSSION TITLE
How can we calculate multiple quantiles simultaneously?

### DISCUSSION THREAD

**Discussion by yyang97** [ott-jax/ott#373] [Q&A]
```
x_test = jax.random.normal(key,shape = (100,))
ott.tools.soft_sort.quantile(x_test, level=0.2)
```
It is good to calculate the 0.2 quantile value of x_test. 

However, how can I simultaneously calculate 0.2 and 0.8 quantile values of x_test?
like

```
x_test = jax.random.normal(key,shape = (100,))
ott.tools.soft_sort.quantile(x_test, level=jnp.array([.2,.8]))
```
However, it reports some errors:
ValueError: All input arrays must have the same shape.

**Comment by michalk8** [comment]
Hi @yyang97 ,

yes, you can [vmap](https://jax.readthedocs.io/en/latest/_autosummary/jax.vmap.html) over the levels, e.g.:
```python
import jax
from ott.tools import soft_sort

key = jax.random.PRNGKey(0)
x_test = jax.random.normal(key, shape=(100,))
levels = jnp.array([0.2, 0.8])

jax.vmap(lambda level: soft_sort.quantile(x_test, level=level))(levels)
# Array([[-0.7245259],
#        [ 0.93384  ]], dtype=float32)
```

**Reply by yyang97** [reply]
Thank you very much for your help! 
May I ask one more question?
How do I calculate jax.vmap(lambda level: soft_sort.quantile(x_test[x_test<1], level=level))(levels), but in a jitted-compilable way?
So I want to calculate the quantiles of the parts lower than 1...

**Comment by marcocuturi** [comment]
another approach would be to have `t+2` targets where `t` is the number of quantiles, and then "craft" more intelligently a target measure. This is what I was referring to in this [TODO](https://github.com/ott-jax/ott/blob/00bbd74091b7c4b35e19e2e09201a3de1741a76c/src/ott/tools/soft_sort.py#LL253C6-L253C6), and which was done in a previous implementation

https://github.com/google-research/google-research/blob/fa49cd0231954458e2b02278d95528b70245f06d/soft_sort/ops.py#L210

If you need this for an application, can you raise an issue? I will try to implement this shortly.

**Reply by marcocuturi** [reply]
@yyang97 you can take a look at https://github.com/ott-jax/ott/pull/382

**Reply by yyang97** [reply]
Thank you Marco for doing this part!

May I ask one more question?
How do I calculate jax.vmap(lambda level: soft_sort.quantile(x_test[x_test<1], level=level))(levels), but in a jitted-compilable way?
So I want to calculate the quantiles of the parts lower than 1...

**Reply by marcocuturi** [reply]
The selection step `x_test[x_test<1]` is going to return an array of variable size. JIT won't like that indeed. Maybe an alternative is only to keep a subset of a predefined size. Another option would be to add padding, to always get the same size, and then set the `weights` parametera to 0 inside soft-sorting for those elements you''d like to discard
